### ISSUE TITLE
Performance issues with c3js

### ISSUE THREAD

**Issue by panthony** [c3js/c3#2303]
I'm currently struggling with the poor performance of c3js as others do and there is a lot of reasons for that.

- displaying a single line chart (from `simple.html`) calls  `getBoundingClientRect` 99 times which has a significant cost ( #1645 )

- displaying a single line chart will also append and remove from DOM a dummy (to mesure font size) a dozen times per chart (probably because we call updateSizes N times per display)

- it creates and resizes absolutely everything, although there is a single line chart it also create, append and resize the nodes for regions, arcs, titles, grid, the Y2 axis and the X axis for subchart (#1591)

- each axis is created up to 5 times at render time, meaning for our line chart with 2 real axes we create in reality (with uncached sizing since it's a new instance etc.) 20 axes because we call updateScales/updateSizes multiple times at render time:

  - initWithData -> redraw -> updateDimension -> updateScales
  - initWithData -> updateDimension -> updateScales
  - initWithData -> initLegend -> updateLegendWithDefaults -> updateLegend -> updateScales
  - initWithData -> initLegend -> updateLegendWithDefaults -> updateLegend -> updateSizes
  - initWithData -> updateScales 

Basically the browser is taking all its time to modify dom, force relayout, mesure values, modify dom, force relayout, etc.. hundreds of times. (#1067)

There was attempt to cache values with `withoutCompute` for some parts but more often than not this is not used because the boolean is not passed.

I started trying to opt-out from unused stuff (starting with axes) and I could reduce the 99 calls to `getBoundingClientRect` to 28 which feels much better on big dashboards but I'm not 100% familiar with the code stack so I can't be sure I'm not breaking some edge cases here and there.

Is there anyone with some knowledge of the code stack that would be interested in trying to improve the performance? Is it worth it to try and do thing nicely and create PRs to this repository?

**Comment by panthony** [comment]
@conceptualspace I removed the `withoutRecompute` and made it the default behavior, the cache is manually cleared when data is (re)loaded (https://github.com/cogniteev/c3/commit/48c7f5bf5ff276169167390d64a64268cc261977).

Also, I opt-out a bunch of stuff when not used like subcharts (https://github.com/cogniteev/c3/commit/33b7a85e23ed265ecca28137e72a7782a5e353df) or y2Axis (https://github.com/cogniteev/c3/commit/800164bfa7b22f0128c068dd4f5fb67670f5e033) which save a massing amount of works.

The thing is that I do not use the API at all to load data so I may have broken part of the API when data is loaded on the fly and add new axis.

**Comment by conceptualspace** [comment]
i hear u.. 

back to topic: there also seems to be quite a performance regression going from c3 v0.4x to v0.5x. seems to have something to do with the transitions 🙍‍♂️
